keps/sig-scheduling/20200612-deducted-topology-manager.md

```diff
@@ -0,0 +1,177 @@
+approvers:
+  - "@huang-wei"
+editor: TBD
+creation-date: 2020-05-28
+last-updated: 2020-05-28
+status: implementable
+see-also:
+superseded-by:
+---
+# Topology Aware Scheduling
+
+<!-- toc -->
+- [Summary](#Summary)
+- [Motivation](#Motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#Proposal)
+- [Use cases](#use-cases)
+- [Test plans](#test-plans)
+- [Graduation criteria](#graduation-criteria)
+- [Implementation history](#implementation-history)
+<!-- /toc -->
+
+# Summary
+
+This document describes behavior of the Kubernetes Scheduler which take into
+account worker node NUMA topology.
+
+# Motivation
+
+After Topology Manager was introduced the problem of launching pod in the
+cluster where worker nodes have different NUMA topology and different amount
+of resources in that topology became actual. Pod could be scheduled on the node
+where total amount of resources are enough, but resource distribution could not
+satisfy the appropriate Topology policy. In this case the pod failed to start. Much
+better behaviour for scheduler would be to select appropriate node where admit
+handlers may pass.
+
+
+## Goals
+
+-   Make scheduling process more precise when we have NUMA topology on the
+worker node.
+
+## Non-Goals
+
+-   Do not change Topology Manager behaviour to be able to work with policy in
+the PodSpec
+-   This Proposal requires exposing NUMA topology information. This KEP doesn't
+describe how to expose all necessary information it just declare what kind of
+information is necessary.
+
+# Proposal
+Kube-scheduler builtin plugin will be added to the main tree. This plugin
+implements a deducted version of TopologyManager it’s different from original topology manager algorithm.
+Plugin checks the ability to run pod only in case of single-numa-node policy on the
+node, since it is the most strict policy, it implies that the launch on the node with
+other existing policies will be successful if the condition for single-numa-node policy passed for the worker node.
+Proposed plugin will use node label to identify which topology policy is
+enabled on the node.
+To work, this plugin requires topology information of the available resource on the worker nodes.
+
+
+## Node labels
+
+Node label contains the name of the topology policy currently implemented in kubelet.
+
+Proposed Node Label may look like this:
+  `beta.kubernetes.io/topology=none|best-effort|restricted|single-numa-node`
+
+It is based on [this Proposal] (https://github.com/kubernetes/enhancements/pull/1340)
+
+To use these labels both in kube-scheduler and in kubelet string constants of these labels should be moved from pkg/kubelet/cm/topologymanager/ and pkg/kubelet/apis/config/types.go to
+pkg/apis/core/types.go
+
+## Topology format
+
+Available resources with topology of the node should be stored in CRD. Format of the topology described
+[in this document](https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit).
```

Comment:
who is going to maintain this CRD? sig-node?

Reply:
CRD was suggested as one of the approach during discussion in sig-node meeting. And interested persons (from RedHat) were involved to discussion of CRD format. Since we plan to create/update CRD not directly from kubelet, but from separate daemon (implemented as daemon set), I think the authors of it will maintain CRD too, including me.

Reply:
Evolving the API spec from a CRD is a good starting point, at some time when the spec is mature, we can merge it back to a core API. RuntimeClass is a good example - it incubated as a CRD in alpha phase, and then merged back into upstream in the beta phase.